File: cpa-ios-demo/Sources/DomainViewController.m
//
//  Copyright (c) European Broadcasting Union. All rights reserved.
//
//  Licence information is available from the LICENCE file.
//

#import "DomainViewController.h"

#import "CPAProvider.h"

NSString *NameForDomain(NSString *domain)
{
    static NSDictionary *s_names = nil;
    static dispatch_once_t s_onceToken;
    dispatch_once(&s_onceToken, ^{
        s_names = @{ @"playlist.rts.ch" : NSLocalizedString(@"RTS Playlist", nil),
                     @"hbbtv.rts.ch" : NSLocalizedString(@"RTS HbbTV", nil),
                     @"unsupported.ch" : NSLocalizedString(@"Unsupported domain", nil) };
    });
    return s_names[domain] ?: NSLocalizedString(@"Unknown domain", nil);
}

@interface DomainViewController ()

@property (nonatomic, weak) IBOutlet UILabel *tokenLabel;
@property (nonatomic, weak) IBOutlet UISwitch *userTokenSwitch;
@property (nonatomic, weak) IBOutlet UISwitch *forceRenewalSwitch;
@property (nonatomic, weak) IBOutlet UISwitch *customTransitionSwitch;

@end

@implementation DomainViewController

#pragma mark Accessors and mutators

- (void)setDomain:(NSString *)domain
{
    _domain = domain;
    
    self.title = [NSString stringWithFormat:@"%@ (%@)", NameForDomain(domain), domain];
}

#pragma mark View lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self reloadData];
}

#pragma mark UI

- (void)reloadData
{
    CPAToken *token = [[CPAProvider defaultProvider] tokenForDomain:self.domain];
    if (token) {
        self.tokenLabel.text = [NSString stringWithFormat:@"%@\n(%@)", token.value, (token.type == CPATokenTypeClient) ? NSLocalizedString(@"Client", nil) : NSLocalizedString(@"User", nil)];
    }
    else {
        self.tokenLabel.text = NSLocalizedString(@"None", nil);
    }
}

#pragma mark Token retrieval

- (IBAction)retrieveToken:(id)sender
{
    CPAToken *existingToken = [[CPAProvider defaultProvider] tokenForDomain:self.domain];
    if (existingToken && ! self.forceRenewalSwitch.on) {
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"Information", nil)
                                                            message:NSLocalizedString(@"A token is already available", nil)
                                                           delegate:nil
                                                  cancelButtonTitle:NSLocalizedString(@"Dismiss", nil)
                                                  otherButtonTitles:nil];
        [alertView show];
        return;
    }
    
    CPACredentialsPresentationBlock credentialsPresentationBlock = nil;
    if (self.customTransitionSwitch.on) {
        credentialsPresentationBlock = ^(UIViewController *viewController, CPAPresentationAction action) {
            if (action == CPAPresentationActionShow) {
                [self.navigationController pushViewController:viewController animated:YES];
            }
            else {
                [self.navigationController popViewControllerAnimated:YES];
            }
        };
    }
    
    CPATokenType type = self.userTokenSwitch.on ? CPATokenTypeUser : CPATokenTypeClient;
    [[CPAProvider defaultProvider] requestTokenForDomain:self.domain withType:type credentialsPresentationBlock:credentialsPresentationBlock completionBlock:^(CPAToken *token, NSError *error) {
        if (error) {
            UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"Error", nil)
                                                                message:[error localizedDescription]
                                                               delegate:nil
                                                      cancelButtonTitle:NSLocalizedString(@"Dismiss", nil)
                                                      otherButtonTitles:nil];
            [alertView show];
            return;
        }
        
        [self reloadData];
    }];
}

@end

File: cpa-ios-demo/Sources_Swift/DomainViewController.swift
//
//  Copyright (c) European Broadcasting Union. All rights reserved.
//
//  License information is available from the LICENSE file.
//

import UIKit

private func nameForDomain(domain: String!) -> String? {
    struct Static {
        static let names = [
            "playlist.rts.ch" : NSLocalizedString("RTSPlaylist", comment: ""),
            "hbbtv.rts.ch" : NSLocalizedString("TS HbbTV", comment: ""),
            "unsupported.ch" : NSLocalizedString("Unsupported domain", comment: "")
        ]
    }
    return Static.names[domain]
}

class DomainViewController: UIViewController {
    
    // MARK: Accessors and mutators
    
    var domain: String! {
        didSet {
            if let name = nameForDomain(domain) {
                self.title = "\(name) (\(domain))"
            }
            else {
                self.title = nil
            }
        }
    }
    
    @IBOutlet private weak var tokenLabel: UILabel!
    @IBOutlet private weak var userTokenSwitch: UISwitch!
    @IBOutlet private weak var forceRenewalSwitch: UISwitch!
    @IBOutlet private weak var customTransitionSwitch: UISwitch!
    
    // MARK: View lifecycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.reloadData()
    }
    
    // MARK: UI
    
    func reloadData() {
        if let token = CPAProvider.defaultProvider()?.tokenForDomain(self.domain) {
            let tokenTypeString = (token.type == CPATokenType.Client) ? NSLocalizedString("Client", comment: "") : NSLocalizedString("User", comment: "")
            self.tokenLabel.text = "\(token.value)\n(\(tokenTypeString))"
        }
        else {
            self.tokenLabel.text = NSLocalizedString("None", comment: "")
        }
    }
    
    // MARK: Actions
    
    @IBAction func retrieveToken(sender: UIButton) {
        let existingToken = CPAProvider.defaultProvider()?.tokenForDomain(self.domain)
        if (existingToken != nil && !self.forceRenewalSwitch.on) {
            let alertView = UIAlertView(title: NSLocalizedString("Information", comment: ""), message: NSLocalizedString("A token is already available", comment: ""), delegate: nil, cancelButtonTitle: NSLocalizedString("Dismiss", comment: ""))
            alertView.show()
            return
        }
        
        var credentialsPresentationBlock: CPACredentialsPresentationBlock? = nil
        if (self.customTransitionSwitch.on) {
            credentialsPresentationBlock = { (viewController, action) -> Void in
                if (action == CPAPresentationAction.Show) {
                    self.navigationController?.pushViewController(viewController, animated: true)
                }
                else {
                    self.navigationController?.popViewControllerAnimated(true)
                }
            }
        }
        
        let type = self.userTokenSwitch.on ? CPATokenType.User : CPATokenType.Client
        CPAProvider.defaultProvider()?.requestTokenForDomain(domain, withType: type, credentialsPresentationBlock: credentialsPresentationBlock, completionBlock: { (token, error) -> Void in
            if (error != nil) {
                let alertView = UIAlertView(title: NSLocalizedString("Error", comment: ""), message: error?.localizedDescription, delegate: nil, cancelButtonTitle: NSLocalizedString("Dismiss", comment: ""))
                alertView.show()
                return
            }
            
            self.reloadData()
        })
    }
}
